# Introduction to Python

**Prepared by:**

Joaquin A. Castañeda

**Sections:**

1.   Python Basics
2.   Python Lists
3.   Functions and Packages
4.   NumPy
5.   Application

**External Links:**

https://projecteuler.net/problem=1

# Python Basics

In [ ]:
# Hit run code to see the output!
print(5 / 8)

0.625


- `5 / 8` divides 5 by 8.

- `print()` shows the answer.

- Output: `0.625`.

In [ ]:
# Addition and division
print(4 + 5)
print(10 / 2)

# Subtraction
print(5 - 5)

# Multiplication
print(3*5)

9
5.0
0
15


- `+` is for addition.

- `/` is for division.

- `-` is for subtraction.

- `*` is for multiplication.

In [ ]:
# Create a variable saving

savings = 100

# Print out savings

print(savings)

100


- `savings = 100` saves 100.

- `print(savings)` shows it.

In [ ]:
# Create the variables monthly_savings and num_months

monthly_savings = 10
num_months = 4

# Multiply monthly_savings and num_months
new_savings = monthly_savings*num_months

# Print new_savings
print(new_savings)

40


- `monthly_savings = 10` saves 10 per month.

- `num_months = 4` sets the number of months.

- `monthly_savings * num_months` finds total savings.

- `new_savings` stores the result.

- `print(new_savings)` shows it.

In [ ]:
# Create a variable half
half = .5

# Create a variable intro
intro = "Hello! How are you?"

# Create a variable is_good
is_good = True

- `half = .5` stores a decimal.

- `intro = "Hello! How are you?"` stores a text.

- `is_good = True` stores a true/false value.

In [ ]:
savings = 100
new_savings = 40

# Calculate total_savings using savings and new_savings

total_savings = savings + new_savings


print(total_savings)

# Print the type of total_savings
print(type(total_savings))

140
<class 'int'>


- `total_savings = savings + new_savings` adds the two amounts.

- `print(total_savings)` shows the total.

- `print(type(total_savings))` shows the data type (int).

# Python Lists

In [ ]:
hall = 11.25
kit = 18.0
liv = 20.0
bed = 10.75
bath = 9.50

# Create list areas

areas = [hall, kit, liv, bed, bath]

# Print areas

print(areas)


[11.25, 18.0, 20.0, 10.75, 9.5]


- `areas = [hall, kit, liv, bed, bath]` makes a list of room sizes.

- `print(areas)` shows the list.

In [ ]:
# House information as list of lists
house = [["hallway", hall],
         ["kitchen", kit],
         ["living room", liv],
         ['bedroom', bed],
         ['bathroom', bath]]

# Print out house
print(house)

[['hallway', 11.25], ['kitchen', 18.0], ['living room', 20.0], ['bedroom', 10.75], ['bathroom', 9.5]]


- `house = [[...], ...]` makes a list of lists with room names and sizes.

- `print(house)` shows the full structure.

In [ ]:
# Create the areas list
areas = ["hallway", 11.25, "kitchen", 18.0, "living room", 20.0, "bedroom", 10.75, "bathroom", 9.50]

# Print out second element from areas
print(areas[1])

# Print out last element from areas
print(areas[-1])

# Print out the area of the living room
print(areas[5])

11.25
9.5
20.0


- `areas[1]` gets the second item (hallway size).

- `areas[-1]` gets the last item (bathroom size).

- `areas[5]` gets the living room size.

In [ ]:
# Create the areas list
areas = ["hallway", 11.25, "kitchen", 18.0, "living room", 20.0, "bedroom", 10.75, "bathroom", 9.50]

# Use slicing to create downstairs
downstairs = areas[:6]

# Use slicing to create upstairs
upstairs = areas[-4:]

# Print out downstairs and upstairs
print(downstairs)
print(upstairs)

['hallway', 11.25, 'kitchen', 18.0, 'living room', 20.0]
['bedroom', 10.75, 'bathroom', 9.5]


- `areas[:6]` gets the first 6 items for `downstairs`.

- `areas[-4:]` gets the last 4 items for `upstairs`.

- `print(downstairs)` shows downstairs rooms.

- `print(upstairs)` shows upstairs rooms.

In [ ]:
house = [["hallway", 11.25],
         ["kitchen", 18.0],
         ["living room", 20.0],
         ["bedroom", 10.75],
         ["bathroom", 9.50]]

# Subset the house list
house[4][1]

9.5

- `house[4][1]` gets the size of the fifth room (bathroom).

In [ ]:
# Create the areas list
areas = ["hallway", 11.25, "kitchen", 18.0, "living room", 20.0, "bedroom", 10.75, "bathroom", 9.50]

# Correct the bathroom area

areas[-1] = 10.50

# Change "living room" to "chill zone"

areas[4] = "chill zone"


- `areas[-1] = 10.50` updates the bathroom size.

- `areas[4] = "chill zone"` renames the living room.

In [ ]:
# Create the areas list and make some changes
areas = ["hallway", 11.25, "kitchen", 18.0, "chill zone", 20.0,
         "bedroom", 10.75, "bathroom", 10.50]

# Add poolhouse data to areas, new list is areas_1
areas_1 = areas + ['poolhouse', 24.5]

# Add garage data to areas_1, new list is areas_2
areas_2 = areas_1 + ['garage', 15.45]

- `areas + ['poolhouse', 24.5]` adds the poolhouse to create `areas_1`.

- `areas_1 + ['garage', 15.45]` adds the garage to create `areas_2`.

In [ ]:
# Delete the poolhouse items from the list

del areas[10:12]

# Print the updated list

print(areas)

['hallway', 11.25, 'kitchen', 18.0, 'chill zone', 20.0, 'bedroom', 10.75, 'bathroom', 10.5]


- `del areas[10:12]` removes the poolhouse name and size.

- `print(areas)` shows the updated list.

In [ ]:
# Create list areas
areas = [11.25, 18.0, 20.0, 10.75, 9.50]

# Change this command
areas_copy = areas[:]

# Change areas_copy
areas_copy[0] = 5.0

# Print areas
print(areas)

[11.25, 18.0, 20.0, 10.75, 9.5]


- `areas[:]` makes a copy of the list so the original stays unchanged.

- `areas_copy[0] = 5.0` changes the first item in the copy only.

- `print(areas)` shows the original list is still the same.

# Functions and Packages

In [ ]:
# Create variables var1 and var2
var1 = [1, 2, 3, 4]
var2 = True

# Print out type of var1
print(type(var1))

# Print out length of var1
print(len(var1))

# Convert var2 to an integer: out2
out2 = int(var2)

<class 'list'>
4


- `var1 = [1, 2, 3, 4]` creates a list.

- `var2 = True` creates a boolean variable.

- `type(var1)` shows that `var1` is a list.

- `len(var1)` gives the number of items in `var1`.

- `int(var2)` converts `True` to `1`.

In [ ]:
# To get help on the max() function, use either:

'''

help(max)

or

?max

'''

'\n\nhelp(max)\n\nor\n\n?max\n\n'

- `help(max)` shows Python’s documentation for the `max` function in the console.

- `?max` works in Jupyter/Colab to display a quick info popup about `max`.

- Use `help()` in scripts; `?` is for interactive notebooks only.

In [ ]:
# Create lists first and second
first = [11.25, 18.0, 20.0]
second = [10.75, 9.50]

# Paste together first and second: full
full = first + second

# Sort full in descending order: full_sorted
full_sorted = sorted(full, reverse=True)

# Print out full_sorted
print(full_sorted)

[20.0, 18.0, 11.25, 10.75, 9.5]


- `first + second` combines the two lists into `full`.

- `sorted(full, reverse=True)` sorts the list from largest to smallest.

- `print(full_sorted)` shows the sorted list.

In [ ]:
# string to experiment with: place
place = "poolhouse"

# Use upper() on place
place_up = place.upper()

# Print out place and place_up
print(place)
print(place_up)

# Print out the number of o's in place
print(place.count('o'))

poolhouse
POOLHOUSE
3


- `place.upper()` makes all letters uppercase and stores in `place_up`.

- `print(place)` shows the original string.

- `print(place_up)` shows the uppercase version.

- `place.count('o')` counts how many times 'o' appears in the string.

In [ ]:
# Create list areas
areas = [11.25, 18.0, 20.0, 10.75, 9.50]

# Print out the index of the element 20.0
print(areas.index(20.0))

# Print out how often 9.50 appears in areas
print(areas.count(9.50))

2
1


- `areas.index(20.0)` finds the position of 20.0 in the list.

- `areas.count(9.50)` counts how many times 9.50 appears.

In [ ]:
# Create list areas
areas = [11.25, 18.0, 20.0, 10.75, 9.50]

# Use append twice to add poolhouse and garage size
areas.append(24.5)
areas.append(15.45)

# Print out areas
print(areas)

# Reverse the orders of the elements in areas
areas.reverse()

# Print out areas
print(areas)

[11.25, 18.0, 20.0, 10.75, 9.5, 24.5, 15.45]
[15.45, 24.5, 9.5, 10.75, 20.0, 18.0, 11.25]


- `areas.append(24.5)` adds 24.5 (poolhouse) to the end of the list.

- `areas.append(15.45)` adds 15.45 (garage) to the end of the list.

- `print(areas)` shows the list with the new elements.

- `areas.reverse()` flips the order of elements in the list.

- `print(areas)` shows the reversed list.

In [ ]:
# Import the math package
import math as m

# Calculate C
C = 2 * 0.43 * m.pi

# Calculate A
A = m.pi * 0.43 ** 2

print("Circumference: " + str(C))
print("Area: " + str(A))

Circumference: 2.701769682087222
Area: 0.5808804816487527


- `import math as m` imports the math module with shortcut `m`.

- `C = 2 * 0.43 * m.pi` calculates the circumference of a circle with radius 0.43.

- `A = m.pi * 0.43 ** 2` calculates the area of the same circle.

- `str(C)` and `str(A)` convert numbers to strings for printing.

- `print("Circumference: " + str(C))` shows the circumference.

- `print("Area: " + str(A))` shows the area.

In [ ]:
# Import pi function of math package
from math import pi

# Calculate C
C = 2 * 0.43 * pi

# Calculate A
A = pi * 0.43 ** 2

print("Circumference: " + str(C))
print("Area: " + str(A))

Circumference: 2.701769682087222
Area: 0.5808804816487527


- `from math import pi` imports only `pi` from the math module.

- `C = 2 * 0.43 * pi` calculates the circle’s circumference.

- `A = pi * 0.43 ** 2` calculates the circle’s area.

- `str()` converts numbers to strings for printing.

- `print("Circumference: " + str(C))` shows the circumference.

- `print("Area: " + str(A))` shows the area.

# NumPy

In [ ]:
# Import the numpy package as np
import numpy as np

baseball = [180, 215, 210, 210, 188, 176, 209, 200]

# Create a numpy array from baseball: np_baseball

np_baseball = np.array(baseball)

# Print out type of np_baseball

print(type(np_baseball))


<class 'numpy.ndarray'>


- `import numpy as np` imports the NumPy library with shortcut `np`.

- `np.array(baseball)` converts the Python list `baseball` into a NumPy array `np_baseball`.

- `type(np_baseball)` shows that `np_baseball` is now a NumPy array (`<class 'numpy.ndarray'>`).

In [ ]:
# Create a numpy array from height_in: np_height_in

np_height_in = np.array(height_in)

# Print out np_height_in

print(np_height_in)

# Convert np_height_in to m: np_height_m

np_height_m = np_height_in * 0.0254

# Print np_height_m

print(np_height_m)


NameError: name 'height_in' is not defined

- `np.array(height_in)` converts the list `height_in` to a NumPy array `np_height_in`.

- `print(np_height_in)` shows the array in inches.

- `np_height_in * 0.0254` converts inches to meters and stores in `np_height_m`.

- `print(np_height_m)` shows the heights in meters.

In [ ]:
np_weight_lb = np.array(weight_lb)
np_height_in = np.array(height_in)

# Print out the weight at index 50

print(np_weight_lb[50])

# Print out sub-array of np_height_in: index 100 up to and including index 110

print(np_height_in[100:111])


- `np.array(weight_lb)` and `np.array(height_in)` convert the lists to NumPy arrays.

- `np_weight_lb[50]` gets the weight at index 50.

- `np_height_in[100:111]` gets a sub-array from index 100 to 110 (slicing excludes the last index, so we use 111).

- `print()` displays the values.

In [ ]:
baseball = [[180, 78.4],
            [215, 102.7],
            [210, 98.5],
            [188, 75.2]]

# Create a 2D numpy array from baseball: np_baseball

np_baseball = np.array(baseball)

# Print out the type of np_baseball

print(type(np_baseball))

# Print out the shape of np_baseball

print(np_baseball.shape)


- `np.array(baseball)` converts the 2D list `baseball` into a 2D NumPy array `np_baseball`.

- `type(np_baseball)` shows that `np_baseball` is a NumPy array (`<class 'numpy.ndarray'>`).

- `np_baseball.shape` returns the dimensions of the array (rows, columns).

- `print()` displays the type and shape.

In [ ]:
# Create a 2D numpy array from baseball: np_baseball
np_baseball = np.array(baseball)

# Print out the shape of np_baseball

print(np_baseball.shape)

- `np.array(baseball)` makes a 2D NumPy array `np_baseball`.

- `np_baseball.shape` shows its dimensions (rows, columns).

- `print()` displays the shape.

In [ ]:
np_baseball = np.array(baseball)

# Print out the 50th row of np_baseball

print(np_baseball[49, :])

# Select the entire second column of np_baseball: np_weight_lb

np_weight_lb = np_baseball[ : , 1]

# Print out height of 124th player

print(np_baseball[123])

- `np_baseball[49, :]` selects the 50th row (all columns) of the array.

- `np_baseball[:, 1]` selects the entire second column and stores it in `np_weight_lb`.

- `np_baseball[123]` gets all data (height and weight) of the 124th player.

- `print()` displays the selected rows or columns.

In [ ]:
np_baseball = np.array(baseball)

# Print out addition of np_baseball and updated

print(np_baseball + updated)

# Create numpy array: conversion

conversion = np.array([0.0254, 0.453592, 1])

# Print out product of np_baseball and conversion

print(np_baseball * conversion)

- `np_baseball + updated` adds corresponding elements of `np_baseball` and `updated` arrays.

- `conversion = np.array([0.0254, 0.453592, 1])` creates an array of conversion factors (inches→meters, pounds→kg, and 1 for no change).

- `np_baseball * conversion` multiplies each column of `np_baseball` by the corresponding conversion factor.

- `print()` shows the results of addition and multiplication.

In [ ]:
# Create np_height_in from np_baseball

np_height_in = np_baseball[:,0]

# Print out the mean of np_height_in

print(np.mean(np_height_in))

# Print out the median of np_height_in

print(np.median(np_height_in))


- `np_baseball[:,0]` selects the first column and stores in `np_height_in`.

- `np.mean(np_height_in)` calculates the average height.

- `np.median(np_height_in)` calculates the median height.

- `print()` displays the mean and median values.

In [ ]:
avg = np.mean(np_baseball[:,0])
print("Average: " + str(avg))

# Print median height
med = np.median(np_baseball[:,0])
print("Median: " + str(med))

# Print out the standard deviation on height
stddev = np.std(np_baseball[:,0])
print("Standard Deviation: " + str(stddev))

# Print out correlation between first and second column
corr = np.corrcoef(np_baseball[:, 0], np_baseball[:, 1])
print("Correlation: " + str(corr))

- `np.mean(np_baseball[:,0])` calculates the average height.

- `np.median(np_baseball[:,0])` calculates the median height.

- `np.std(np_baseball[:,0])` calculates the standard deviation of heights.

- `np.corrcoef(np_baseball[:,0], np_baseball[:,1])` calculates the correlation between height and weight.

- `print()` displays all the results.

# Application

*This challenge is the first item in the archives section of ProjectEuler.net*

Multiples of 3 or 5:


If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23.
Find the sum of all the multiples of 3 or 5 below 1000.

In [4]:
# We need a code that determines divisibility for 3 and 5; for that we use the modulo operator.
# We need a code that scans natrual numbers from 0 to 999; for that we use a for loop.
# Lets us try an if-else syntax for filtering what we need.
# We are asked for the sum, hence we add.

# Let us put everything in an empty list named Liszt
liszt = []

for i in range(1000):
  if i%3 == 0 or i%5 == 0:
    liszt.append(i)

print(sum(liszt))

233168


- `liszt = []` creates an empty list to store the multiples.
- `range(1000)` generates numbers from 0 to 999, which are all natural numbers below 1000.
- `i % 3 == 0` checks if the number is divisible by 3.
- `i % 5 == 0` checks if the number is divisible by 5.
- `or` combines the two conditions so that numbers divisible by either 3 or 5 are included.
- `liszt.append(i)` adds the number to the list when the condition is true.
- `sum(liszt)` calculates the total sum of all multiples in the list.
- `print()` displays the final sum, which is 233168.

**End of Notebook.**